# Task 4.9 Part 1

## Step 3 Import Libraries & Data

#### Import Libraries

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

#### Import Data

In [2]:
# Create folder path
path = r'/Users/madelinegamache/Documents/InstaCart Basket Analysis'

In [3]:
# Import customers.csv file
df_customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data','customers.csv'), index_col = False)

In [4]:
# Check output
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [5]:
# Import orders_products_grouped_new 
ords_prods_grouped = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_grouped_new.pkl'))

In [6]:
# Check output
ords_prods_grouped.head()

order_id  user_id  order_number  weekday_order_placed  order_hour_of_day  \
0   2539329        1             1                     2                  8   
1   2398795        1             2                     3                  7   
2    473747        1             3                     3                 12   
3   2254736        1             4                     4                  7   
4    431534        1             5                     4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...    price_range_loc     busiest_day     busiest_days  \
0         Soda  ...  Mid-range product  Regularly busy   Regularly busy   
1         Soda  ...  Mid-range product  Regularly busy  Least busy days   
2         Soda  ...  Mid-range product  Regularly busy  Least busy days   
3         Soda  ...  Mid-range product      Least busy  Least busy days   
4         Soda  ...  Mid-range product      Least busy  Least busy days   

  busiest_period_of_day max_order  loyalty_flag mean_price spending_flag  \
0        Average orders        10  New customer   6.367797   Low spender   
1        Average orders        10  New customer   6.367797   Low spender   
2           Most orders        10  New customer   6.367797   Low spender   
3        Average orders        10  New customer   6.367797   Low spender   
4           Most orders        10  New customer   6.367797   Low spender   

  median_prior_orders         frequency_flag  
0                20.5  Non-frequent customer  
1                20.5  Non-frequent customer  
2                20.5  Non-frequent customer  
3                20.5  Non-frequent customer  
4                20.5  Non-frequent customer  

[5 rows x 25 columns]

## Step 4 Wrangle Data

In [8]:
#Checking columns listed in customers.csv dataframe for unnecessary or illogical names
df_customers.columns

Index(['user_id', 'First Name', 'Surnam', 'Gender', 'STATE', 'Age',
       'date_joined', 'n_dependants', 'fam_status', 'income'],
      dtype='object')

In [9]:
# Renaming columns so they're more intuitive
df_customers.rename(columns = {'Surnam': 'Last Name'}, inplace = True)

In [10]:
# Renaming columns so they're more intuitive
df_customers.rename(columns = {'n_dependants': 'Number of Dependants'}, inplace = True)

In [11]:
# Renaming columns so they're more intuitive
df_customers.rename(columns = {'fam_status': 'Marital Status'}, inplace = True)

In [12]:
# Checking new column headers
df_customers.head()

user_id First Name Last Name  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   Number of Dependants Marital Status  income  
0                     3        married  165665  
1                     0         single   59285  
2                     2        married   99568  
3                     0         single   42049  
4                     1        married   40374

In [13]:
# Checking basic info of customers.csv
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   user_id               206209 non-null  int64 
 1   First Name            194950 non-null  object
 2   Last Name             206209 non-null  object
 3   Gender                206209 non-null  object
 4   STATE                 206209 non-null  object
 5   Age                   206209 non-null  int64 
 6   date_joined           206209 non-null  object
 7   Number of Dependants  206209 non-null  int64 
 8   Marital Status        206209 non-null  object
 9   income                206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [14]:
# Checking First name for missing values
df_customers['First Name'].value_counts(dropna = False)

NaN        11259
Marilyn     2213
Barbara     2154
Todd        2113
Jeremy      2104
           ...  
Merry        197
Eugene       197
Garry        191
Ned          186
David        186
Name: First Name, Length: 208, dtype: int64

##### Confirms 11259 missing values in First Name column. This is a small enough percentage to leave blank for now, as it does not make sense to randomly input names and there is no way to determine their names.

## Step 5  Data Quality & Consistency Checks

In [17]:
# Check customers.csv data types
df_customers.dtypes

user_id                  int64
First Name              object
Last Name               object
Gender                  object
STATE                   object
Age                      int64
date_joined             object
Number of Dependants     int64
Marital Status          object
income                   int64
dtype: object

In [18]:
# While user_id is a number, we won't use it for calculations so it should be changed to a "str" variable
df_customers['user_id'] = df_customers['user_id'].astype('str')

In [19]:
# Check output
print(df_customers['user_id'].dtypes)

object


In [20]:
# Checking the descriptive statistics of customers.csv
df_customers.describe()

Age  Number of Dependants         income
count  206209.000000         206209.000000  206209.000000
mean       49.501646              1.499823   94632.852548
std        18.480962              1.118433   42473.786988
min        18.000000              0.000000   25903.000000
25%        33.000000              0.000000   59874.000000
50%        49.000000              1.000000   93547.000000
75%        66.000000              3.000000  124244.000000
max        81.000000              3.000000  593901.000000

In [21]:
# Check for data consistency
for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)
  else:
    print ('No mixed dtypes', col)

No mixed dtypes user_id
First Name
No mixed dtypes Last Name
No mixed dtypes Gender
No mixed dtypes STATE
No mixed dtypes Age
No mixed dtypes date_joined
No mixed dtypes Number of Dependants
No mixed dtypes Marital Status
No mixed dtypes income


In [22]:
# Change the datatype of First Name to a str
df_customers['First Name'] = df_customers['First Name'].astype('str')

In [23]:
# Check for data consistency again
for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)
  else:
    print ('No mixed dtypes', col)

No mixed dtypes user_id
No mixed dtypes First Name
No mixed dtypes Last Name
No mixed dtypes Gender
No mixed dtypes STATE
No mixed dtypes Age
No mixed dtypes date_joined
No mixed dtypes Number of Dependants
No mixed dtypes Marital Status
No mixed dtypes income


In [24]:
# Check for missing values again
df_customers.isnull().sum()

user_id                 0
First Name              0
Last Name               0
Gender                  0
STATE                   0
Age                     0
date_joined             0
Number of Dependants    0
Marital Status          0
income                  0
dtype: int64

##### Now that the first name variable is a 'str' variable it doesn't register the blanks as missing values anymore.

In [25]:
# Check any duplicate entries in customer.csv and assign them to df_dups
df_dups = df_customers[df_customers.duplicated()]

In [26]:
df_dups

Empty DataFrame
Columns: [user_id, First Name, Last Name, Gender, STATE, Age, date_joined, Number of Dependants, Marital Status, income]
Index: []

##### No duplicates found

## Step 6 Combine Data

In [27]:
ords_prods_grouped.shape

(32404859, 25)

In [28]:
# Check datatypes for compatibility with customers.csv
ords_prods_grouped.dtypes

order_id                     int64
user_id                      int64
order_number                 int64
weekday_order_placed         int64
order_hour_of_day            int64
days_since_prior_order     float64
product_id                   int64
add_to_cart_order            int64
reordered                    int64
product_name                object
aisle_id                     int64
department_id                int64
prices                     float64
_merge                    category
 price_range_loc            object
price_range_loc             object
busiest_day                 object
busiest_days                object
busiest_period_of_day       object
max_order                    int64
loyalty_flag                object
mean_price                 float64
spending_flag               object
median_prior_orders        float64
frequency_flag              object
dtype: object

In [29]:
# Changing the user_id dtype in the ords_prods_grouped file into a string variable so it matches with df_customers
ords_prods_grouped['user_id'] = ords_prods_grouped['user_id'].astype('str')

In [30]:
ords_prods_grouped.dtypes

order_id                     int64
user_id                     object
order_number                 int64
weekday_order_placed         int64
order_hour_of_day            int64
days_since_prior_order     float64
product_id                   int64
add_to_cart_order            int64
reordered                    int64
product_name                object
aisle_id                     int64
department_id                int64
prices                     float64
_merge                    category
 price_range_loc            object
price_range_loc             object
busiest_day                 object
busiest_days                object
busiest_period_of_day       object
max_order                    int64
loyalty_flag                object
mean_price                 float64
spending_flag               object
median_prior_orders        float64
frequency_flag              object
dtype: object

In [31]:
# Merging the two dfs based on the user_id column
customers_ords_prods_combined = ords_prods_grouped.merge(df_customers, on = 'user_id')

In [32]:
# Check output
customers_ords_prods_combined.head()

order_id user_id  order_number  weekday_order_placed  order_hour_of_day  \
0   2539329       1             1                     2                  8   
1   2398795       1             2                     3                  7   
2    473747       1             3                     3                 12   
3   2254736       1             4                     4                  7   
4    431534       1             5                     4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...         frequency_flag  First Name  Last Name  Gender  \
0         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
1         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
2         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
3         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
4         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   

     STATE Age date_joined Number of Dependants Marital Status  income  
0  Alabama  31   2/17/2019                    3        married   40423  
1  Alabama  31   2/17/2019                    3        married   40423  
2  Alabama  31   2/17/2019                    3        married   40423  
3  Alabama  31   2/17/2019                    3        married   40423  
4  Alabama  31   2/17/2019                    3        married   40423  

[5 rows x 34 columns]

## Step 7 Cleanup Notebook

In [34]:
# Cleaning up column headers so they are consistent typed in merged dataframe
customers_ords_prods_combined.rename(columns = {'weekday_order_placed' : 'order_day_of_week', 'First Name' : 'first_name', 'Last Name' : 'last_name', 'Gender' : 'gender', 'STATE' : 'state', 'Age' : 'age', 'Number of Dependants' : 'total_dependants', 'Marital Status' : 'marital_status'}, inplace = True)

In [35]:
# Check column headers
customers_ords_prods_combined.head()

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329       1             1                  2                  8   
1   2398795       1             2                  3                  7   
2    473747       1             3                  3                 12   
3   2254736       1             4                  4                  7   
4    431534       1             5                  4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...         frequency_flag  first_name  last_name  gender  \
0         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
1         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
2         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
3         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   
4         Soda  ...  Non-frequent customer       Linda     Nguyen  Female   

     state age date_joined total_dependants marital_status  income  
0  Alabama  31   2/17/2019                3        married   40423  
1  Alabama  31   2/17/2019                3        married   40423  
2  Alabama  31   2/17/2019                3        married   40423  
3  Alabama  31   2/17/2019                3        married   40423  
4  Alabama  31   2/17/2019                3        married   40423  

[5 rows x 34 columns]

##### All column headers are now visually consistent and typed in lower case

## Step 8 Export Dataframe

In [36]:
# Export as pkl
customers_ords_prods_combined.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'customer_ords_prods_merged.pkl'))